### PR TITLE
Proposal for Template Versioning

### PR DESCRIPTION
**Pull request by parauliya**
Signed-off-by: parauliya <aman@infracloud.io>

### CODE REVIEW

File: proposals/0020/README.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+---
+id: 0020
+title: Template Versioning
+status: published
+authors: Aman Parauliya <aman@infracloud.io>
+---
+
+## Summary
+
+As per the proposal [0013], we are planning to induce workflow service as a part of the template which is used to create a workflow. As per the future perspective we may again have the requirement of modifying the existing format of template and in that case we should also provide the support of previous/current format of the template in parallel to the modification of template format/schema.
```
Comment (gauravgahlot) [inline proposals/0020/README.md]: Follow the practice of one statement per line.

Comment (gauravgahlot) [reply proposals/0020/README.md]: Maybe reword:

> As per the future perspective we may again have the requirement of modifying the existing format of template and in that case we should also provide the support of previous/current format of the template in parallel to the modification of template format/schema.

to something like: 

> As we add new features, we may have to modify the existing template format. Therefore, from a future perspective, we should support template versioning so that we can support existing template formats.

File: proposals/0020/README.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+id: 0020
+title: Template Versioning
+status: published
+authors: Aman Parauliya <aman@infracloud.io>
+---
+
+## Summary
+
+As per the proposal [0013], we are planning to induce workflow service as a part of the template which is used to create a workflow. As per the future perspective we may again have the requirement of modifying the existing format of template and in that case we should also provide the support of previous/current format of the template in parallel to the modification of template format/schema.  
+This is to provide backward compatibility once [0013] is implemented.
```
Comment (gauravgahlot) [inline proposals/0020/README.md]: I think it should be done before we implement 0013.

File: proposals/0020/README.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+Current format or schema of the template should be compatible as well as with the new one.
+
+### Not Goals
+
+Current template schema should be depricated.
+
+## Content
+
+This feature will make the life of existing user a bit easier because he/she doesn't need to migrate to new template schema since we have the support of existing and user will also have the flexibility of using the kind of template he/she would like to use or in another words whichever template schema is fit for his/her usecase.
```
Comment (gauravgahlot) [inline proposals/0020/README.md]: change 
> support of existing

to 

> support for existing templates


The sentence is way too long. Please break it to smaller ones.

**Review by gauravgahlot — CHANGES_REQUESTED** [review]

File: proposals/0020/README.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
+---
+id: 0020
+title: Template Versioning
+status: published
```
Comment (gauravgahlot) [inline proposals/0020/README.md]: Status should be "ideation".

File: proposals/0020/README.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+## Summary
+
+As per the proposal [0013], we are planning to induce workflow service as a part of the template which is used to create a workflow. As per the future perspective we may again have the requirement of modifying the existing format of template and in that case we should also provide the support of previous/current format of the template in parallel to the modification of template format/schema.  
+This is to provide backward compatibility once [0013] is implemented.
+
+## Goals and not Goals
+
+### Goals
+
+Current format or schema of the template should be compatible as well as with the new one.
```
Comment (gauravgahlot) [inline proposals/0020/README.md]: IMO goal should mention versioning.

File: proposals/0020/README.md (hunk 2 above)
Comment (parauliya) [reply proposals/0020/README.md]: Yes, Actually , let's make it this way that this needs to be done in parallel to 0013 or as a part of 0013.

File: proposals/0020/README.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        Tasks         []Task `yaml:"tasks"`
+}
+
+```
+
+Now if we would like to support both of the above templates then we need to identify the struct in which we need to parse the template which we can identify based on the versions.
+
+## Alternatives
+
+We can have the support only for latest template schema and ask user to migrate to the latest format of template schema each time we do a modification in the template schema while providing a stable release. In that case, there will not have the support for the existing schema of template which can affect to the existing users.
```
Comment (gauravgahlot) [inline proposals/0020/README.md]: As another alternative, we can write a utility that can read the old templates and transform them as per the new schema.

**Review by gauravgahlot — CHANGES_REQUESTED** [review]

File: proposals/0020/README.md (hunk 6 above)
Comment (parauliya) [reply proposals/0020/README.md]: Yeah, this can be added.

File: proposals/0020/README.md (hunk 5 above)
Comment (parauliya) [reply proposals/0020/README.md]: Done, refromed the current statement.

File: proposals/0020/README.md (hunk 4 above)
Comment (parauliya) [reply proposals/0020/README.md]: Done.

File: proposals/0020/README.md (hunk 3 above)
Comment (parauliya) [reply proposals/0020/README.md]: Done.

File: proposals/0020/README.md (hunk 6 above)
Comment (parauliya) [reply proposals/0020/README.md]: Done

File: proposals/0020/README.md (hunk 1 above)
Comment (parauliya) [reply proposals/0020/README.md]: Seems good and less complex to me will change.
